app/controllers/v1/sessions_controller.rb

```diff
@@ -138,23 +138,49 @@ def user_login(saml_response)
     end
 
     def render_login(type)
-      login_url, post_params = login_params(type)
+      login_url, @post_params = login_params(type)
       tracker = url_service.tracker
       renderer = ActionController::Base.renderer
       renderer.controller.prepend_view_path(Rails.root.join('lib', 'saml', 'templates'))
       result = renderer.render template: 'sso_post_form',
                                locals: {
                                  url: login_url,
-                                 params: post_params,
+                                 params: @post_params,
                                  id: tracker.uuid,
                                  authn: tracker.payload_attr(:authn_context),
                                  type: tracker.payload_attr(:type)
                                },
                                format: :html
       render body: result, content_type: 'text/html'
+      set_sso_saml_cookie!
       saml_request_stats
     end
 
+    def set_sso_saml_cookie!
+      cookies[Settings.ssoe_eauth_cookie.name] = {
+        value: saml_cookie_content.to_json,
+        expires: nil,
+        secure: Settings.ssoe_eauth_cookie.secure,
+        httponly: true,
+        domain: Settings.ssoe_eauth_cookie.domain
+      }
+    end
+
+    def saml_cookie_content
+      transaction_id = if current_user && url_service.should_uplevel?
+                         JSON.parse(cookies[Settings.ssoe_eauth_cookie.name])['transaction_id']
+                       else
+                         SecureRandom.uuid
+                       end
+
+      {
+        'timestamp' => Time.now.iso8601,
+        'transaction_id' => transaction_id,
+        'saml_request_id' => url_service.tracker&.uuid,
+        'saml_request_query_params' => url_service.query_params
```

Comment:
@ericbuckley does this look right for `saml_request_query_params`?  Any "query_params" we should exclude?